daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala

```diff
@@ -294,197 +294,240 @@ class Runner(
     var machine =
       Speedy.Machine.fromSExpr(script.expr, false, extendedCompiledPackages)
 
-    def stepToValue() = {
+    // Removing the early return only makes this harder to read.
+    @SuppressWarnings(Array("org.wartremover.warts.Return"))
+    def stepToValue(): Future[Unit] = {
```

Comment:
What's the purpose of `stepToValue` returning a future? As I understand it, right now the actual evaluation (`while (!machine.isFinal)`) is synchronous. If it should happen asynchronously then the whole body should be wrapped in `Future { ... }`, correct? If the whole body is wrapped in a `Future`, then `throw` would lead to a failed future.

Reply:
We currently have a lot of code which ends up boiling down to something like `go().flatMap(stepToValue(); …; go())`. If `stepToValue` throws that will not produce a failed future but instead we run into the issue we described. We could wrap `stepToValue` into `Future` at the call site but then we have to do that at every call site which seems much easier to get wrong than changing the type so you’re forced to treat it as a future. Furthermore `Future.failed(…)` also has an advantage over `Future { throw …}` since the latter will execute the code asynchronously on an execution context while the former will just construct the resolved future directly which is much more efficient.

Reply:
> We could wrap `stepToValue` into `Future` at the call site but then we have to do that at every call site which seems much easier to get wrong than changing the type so you’re forced to treat it as a future.

My suggestion was to wrap the implementation of `stepToValue` in `Future { ... }` not the call-site. I.e.
```
def stepToValue(): Future[Unit] = Future {
    ...
}
```

> Furthermore Future.failed(…) also has an advantage over Future { throw …} since the latter will execute the code asynchronously on an execution context while the former will just construct the resolved future directly which is much more efficient.

Sorry, I'm still confused about this. Seeing the type `stepToValue(): Future[Unit]` suggests to me that the stepping is done asynchronously. If the function is supposed to be synchronous, then why not return `Either` or `Try`?

Reply:
I’m not following, you suggest to wrap it in `Future` but then you suggest to return `Either`? Surely we can’t do both? We could do use `Either` but given that we only use it in `Future` contexts we would then have to do the conversion to the failed future at every call site.

Reply:
I'm not suggesting to do both. It's an either or.

The point is, the way I understand `stepToValue` (and `go` seems to largely follow the same pattern) is that it is synchronous code. The `Future` type is only used to handle errors. (Though it's perfectly possible I'm missing something). So, why use `Future`?

Reply:
You are completely right that it’s synchronous. But it’s only used in the context of the `runWithClients` future so it just seemed more convenient to move the conversion into `stepToValue` instead of doing it at the call site. But I don’t feel strongly about this so happy to switch it to `Either` and to the conversion at the call site if you prefer.

Reply:
Okay, wanted to make sure this is intentional. To me this seems confusing, but I'm not sure what Scala best practices say there. Can be fixed in a separate PR, though.

Reply:
After thinking about it for a bit I agree, I’ve changed it to return `Either[RuntimeException, Unit]`. Thanks for the discussion!

Reply:
Thank you!